### PR TITLE
Send system and runtime metrics to the internal metrics data stream

### PR DESCRIPTION
**Pull request by axw**
## Checklist

- [x] Update [CHANGELOG.asciidoc](https://github.com/elastic/apm-server/blob/main/CHANGELOG.asciidoc)
- [x] Update [package changelog.yml](https://github.com/elastic/apm-server/blob/main/apmpackage/apm/changelog.yml) (only if changes to `apmpackage` have been made)
~- [ ] Documentation has been updated~

## How to test these changes

1. Run APM Server
2. Run all opbeans applications, pointed at APM Server (e.g. using apm-integration-testing with `--all-opbeans`)
3. Wait at least 1 minute to ensure all agents have sent some metrics
4. Ensure there are no service-specific application metrics data streams (`metrics-apm.app.*-*`)

## Related issues

Closes https://github.com/elastic/apm-server/issues/7520

### CODE REVIEW

**Comment by simitt** [comment]
@axw is this ready for an initial review already?

**Comment by axw** [comment]
Not quite ready - should be soon.

File: model/modelprocessor/generate_internal_metrics.go [hunk 1]
```diff
@@ -0,0 +1,99 @@
...
+func main() {
+	_, file, _, ok := runtime.Caller(0)
+	if !ok {
+		log.Fatal("runtime.Caller failed")
+	}
+	modelprocessorDir := filepath.Dir(file)
+	repoRoot := filepath.Join(modelprocessorDir, "..", "..")
+
+	internalMetricsFieldsDir := filepath.Join(repoRoot, "apmpackage", "apm", "data_stream", "internal_metrics", "fields")
+	files, err := filepath.Glob(filepath.Join(internalMetricsFieldsDir, "*_metrics.yml"))
```
Comment (marclop) [inline model/modelprocessor/generate_internal_metrics.go:44]: Are we going to keep the `yaml` definitions of the data stream fields in sync with what the agents send in a best-effort manual fashion?

File: model/modelprocessor/datastream.go [hunk 2]
```diff
@@ -84,3 +74,39 @@ func isRUMAgentName(agentName string) bool {
...
+
+	if event.Metricset != nil {
+		// Well-defined system/runtime metrics are stored in the shared
+		// "internal" metrics data stream, as they are not application-specific.
+		//
+		// TODO(axw) agents should indicate that a metricset is internal.
+		// If a metricset is identified as internal, then we'll ignore any
+		// metrics that we don't already know about; otherwise they will end
+		// up creating service-specific data streams.
+		internal := true
```
Comment (marclop) [inline model/modelprocessor/datastream.go:94]: Are there plans to extend the Intake API to accommodate this?

**Review by marclop — APPROVED** [review]
LGTM. Just a question on how to keep these in sync with the agents.

File: model/modelprocessor/generate_internal_metrics.go (hunk 1 above)
Comment (axw) [reply model/modelprocessor/generate_internal_metrics.go:44]: Yes. The metrics change pretty infrequently. Later on we might revert to dynamically mapping metrics identified as being internal, only in the internal metrics data stream.

File: model/modelprocessor/datastream.go (hunk 2 above)
Comment (axw) [reply model/modelprocessor/datastream.go:94]: Only in my head for the moment. I have a recurring reminder to create an issue - I'll try to do this in the next couple of days.

**Comment by stuartnelson3** [comment]
Note: I tested this a couple times with buildsha c00efb988b99c9cc455ac764d5811d0975860e3f, not an official RC. It does contain this change, however.

```
# curl localhost:8200
{
  "build_date": "2022-05-24T00:37:08Z",
  "build_sha": "c00efb988b99c9cc455ac764d5811d0975860e3f",
  "publish_ready": true,
  "version": "8.3.0"
}
```

Reproduction steps:
- start the stack with apm-integration: `./scripts/compose.py start 8.3.0 --remove-orphans`
- add opbeans: `./scripts/compose.py start 8.3.0 --remove-orphans --all-opbeans`
- check that metrics have been ingested in the apm-ui
- navigate to stack management > index management > datastreams, filter for apm-metrics

I observed individual app metric datastreams:
image

**Comment by stuartnelson3** [comment]
confirmed in https://github.com/elastic/apm-server/pull/8264

**Pull request merged**